src/main/java/org/prebid/server/handler/CookieSyncHandler.java

```diff
@@ -127,19 +134,43 @@ public void handle(RoutingContext context) {
 
         gdprService.resultByVendor(GDPR_PURPOSES, vendorIds, gdprAsString, gdprConsent, ip,
                 timeoutFactory.create(defaultTimeout))
-                .setHandler(asyncResult -> handleResult(asyncResult, context, uidsCookie, biddersToSync,
-                        gdprAsString, gdprConsent, cookieSyncRequest.getLimit()));
+                .setHandler(asyncResult -> handleResult(asyncResult, context, uidsCookie, biddersToSync, gdprAsString,
+                        gdprConsent, limit));
     }
 
     /**
      * Returns bidder names to sync.
      * <p>
      * If bidder list was omitted in request, that means sync should be done for all bidders.
      */
-    private Collection<String> biddersToSync(List<String> biddersFromRequest) {
-        return CollectionUtils.isEmpty(biddersFromRequest)
-                ? bidderCatalog.names().stream().filter(bidderCatalog::isActive).collect(Collectors.toSet())
-                : biddersFromRequest;
+    private Collection<String> biddersToSync(List<String> requestBidders, Boolean requestCoop, Integer requestLimit) {
+        if (CollectionUtils.isEmpty(requestBidders)) {
+            return bidderCatalog.names().stream().filter(bidderCatalog::isActive).collect(Collectors.toSet());
+        }
+
+        final boolean coop = requestCoop == null ? defaultCoop : requestCoop;
+
+        if (coop) {
+            final int limit = requestLimit == null ? Integer.MAX_VALUE : requestLimit;
+            final int remaining = limit - requestBidders.size();
+            final Collection<String> coopSync = getCoopSync(requestBidders, remaining);
+            requestBidders.addAll(coopSync);
+        }
+
+        return requestBidders;
+    }
+
+    private Collection<String> getCoopSync(List<String> bidders, int limit) {
+        if (prioritisedBidders == null || limit <= 0) {
+            return Collections.emptyList();
+        }
+
+        return prioritisedBidders.stream()
+                .peek(Collections::shuffle)
```

Comment:
think about it one more, it mb better to shuffle and limit in separate operations.